---
title: "trnF data analysis using DADA2"
author: "Pedro Rodrigues"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 3
    fig_caption: yes

fontsize: 14pt
---

# Introduction

A portion of the _trnF_ gene was sequenced from three batches of burrower bugs collected in peanut fields in southern GA, USA. Here, these three batches of sequencing data, PBB1, PBB2, and PBB3, are quality-trimmed and noise-filtered before being combined and classified into plant taxa.


# Setup

Prepare environment
Load packages

```{r init, warning=FALSE, message=FALSE}

setwd("/Users/ento-user/Documents/Github_Burrower_Bug/Burrower-Bugs-diet-and-pop-gen/ASV_identification")
set.seed("1805")

library(dada2);packageVersion("dada2") # ‘1.30.0’
library(Biostrings); packageVersion("Biostrings") #‘2.70.3’
library(ShortRead); packageVersion("ShortRead") #‘1.60.0’
library(ggplot2); packageVersion("ggplot2") #‘3.5.0’
library(reshape2); packageVersion("reshape2") #‘1.4.4’
library(gridExtra); packageVersion("gridExtra") #‘2.3’
library(phyloseq); packageVersion("phyloseq") #‘1.46.0’
library(dplyr); packageVersion("dplyr") #‘1.1.4’
```

```{r more preparation, warning=FALSE, message=FALSE}
pathPBB1 <- "../diet_data/PBB1"
pathPBB2 <- "../diet_data/PBB2"
pathPBB3 <- "../diet_data/PBB3"
path.out <- "Figures/"
path.rds <- "RDS/"
fnsPBB1 <- list.files(pathPBB1, pattern=".fastq", full.names=TRUE)
fnsPBB2 <- list.files(pathPBB2, pattern=".fastq", full.names=TRUE)
fnsPBB3 <- list.files(pathPBB3, pattern=".fastq", full.names=TRUE)
B49873_e <- "GGTTCAAGTCCCTCTATCCC"
A50272_f <- "ATTTGAACTGGTGACACGAG"
rc <- dada2:::rc
theme_set(theme_bw())
```


# Primer removal and filtering
## Remove Primers and Filter (PBB3)

Remove primers and orient reads:
```{r primers3}
nops3 <- file.path(pathPBB3, "noprimers", basename(fnsPBB3))
prim3 <- removePrimers(fnsPBB3, nops3, primer.fwd=B49873_e, primer.rev=dada2:::rc(A50272_f), orient=TRUE)
```

Inspect length distribution.
```{r length-distro3}
lens.fn3 <- lapply(nops3, function(fn) nchar(getSequences(fn)))
lens3 <- do.call(c, lens.fn3)
hist(lens3, 100)
min(lens3)
max(lens3)
```

There is a wide range of lengths in these samples - this is expected. Let's check what is the minimum size expected based on the trnF database that I created using the NCBI RefSeq chloroplasts genomes

In UNIX, I get the following result (see below)

(base) pd88715@ss-sub3 ref_seq_plastids$ cat trnF_db11Feb22.fa | awk '$0 ~ ">" {if (NR > 1) {print c;} c=0;printf substr($0,2,100) "\t"; } $0 !~ ">" {c+=length($0);} END { print c; }' | cut -f 2 | sort -n | head
101
102
102
102
102
102
102
103
103
103

>NOTE: one of the settings for building this reference trnF intergenic gene database is removing sequences smaller than 100bp

### Filter and Trim PBB3
>Note on maxEE - for fungal identification using ITS1 it was found that increasing maxEE = 8 increased the number of reads per taxa expected to be present in the sample (Rolling et al. 2021); previously I used maxEE=2, and had to rarefy data in downstream analysis to 100 sequences/sample, which was very low; I am using maxEE=8 to increase the number of reads per plant species
source  https://insight.jci.org/articles/view/151663

```{r filter3}
filts3 <- file.path(pathPBB3, "noprimers", "filtered", basename(fnsPBB3))
track3 <- filterAndTrim(nops3, filts3, minQ=3, minLen=100, maxN=0, rm.phix=FALSE, maxEE=8)
track3
```

## Run DADA2

Dereplicate:
```{r derep3, message=FALSE}
drp3 <- derepFastq(filts3, verbose=TRUE)
```


Learn errors:
```{r learn-err3}
err3 <- learnErrors(drp3, errorEstimationFunction=PacBioErrfun, BAND_SIZE=32, multithread=TRUE)
saveRDS(err3, file.path(path.rds, "PBB3_err2.rds")) #name changes here, keeping previous files generated by version 1 of this workflow
```

Inspect errors:
```{r see-err3}
plotErrors(err3)
```

Looks good: base error decreases with higher base quality scores.

Denoise:
```{r dada2-sample3}
dd3 <- dada(drp3, err=err3, BAND_SIZE=32, multithread=TRUE)
saveRDS(dd3, file.path(path.rds, "PBB3_dd3_2.rds"))
```
Not a lot of losses. PacBio stringent pre-filtering probably already excluded most of the low quality sequences.
There is generally more reads than unique reads, which is a good sign.

Read tracking for PBB3:
```{r}
cbind(ccs=prim3[,1], primers=prim3[,2], filtered=track3[,2], denoised=sapply(dd3, function(x) sum(x$denoised)))
```

Sequence table for PBB3:
```{r seqtab3}
st3 <- makeSequenceTable(dd3); dim(st3)
```
Here is a good point to save the sequence table and use this table when merging with other datasets (PBB1 and PBB2).
```{r save seq table3}
saveRDS(st3, "RDS/PBB3_2.rds")
```


------------------------------------------

## Combining data from different runs

>Process the other two datasets (PBB1 and PBB2)

Starting with PBB1, remove primers and orient reads:
```{r primers1, message=FALSE, warning=FALSE}
nops1 <- file.path(pathPBB1, "noprimers", basename(fnsPBB1))
prim1 <- removePrimers(fnsPBB1, nops1, primer.fwd=B49873_e, primer.rev=dada2:::rc(A50272_f), orient=TRUE)
```

Inspect length distribution.
```{r length-distro1}
lens.fn <- lapply(nops1, function(fn) nchar(getSequences(fn)))
lens <- do.call(c, lens.fn)
hist(lens, 100)
min(lens)
max(lens)
```


### Filter and Trim
#### maxEE=8
```{r filter1}
filts1 <- file.path(pathPBB1, "noprimers", "filtered", basename(fnsPBB1))
track1 <- filterAndTrim(nops1, filts1, minQ=3, minLen=100, maxN=0, rm.phix=FALSE, maxEE=8)
track1
```


Dereplicate:
```{r derep1, warning=FALSE, message=FALSE}
drp1 <- derepFastq(filts1, verbose=TRUE)
```

Learn errors:
```{r learn-err1}
err1 <- learnErrors(drp1, errorEstimationFunction=PacBioErrfun, BAND_SIZE=32, multithread=TRUE)
saveRDS(err1, file.path(path.rds, "err_PBB1_2.rds"))
```

Inspect errors:
```{r see-err1}
plotErrors(err1)
```

Looks good.

Denoise:
```{r dada2-sample1}
dd1 <- dada(drp1, err=err1, BAND_SIZE=32, multithread=TRUE)
saveRDS(dd1, file.path(path.rds, "PBB1_dd1_2.rds"))
```

Read tracking for PBB1:
```{r}
cbind(ccs=prim1[,1], primers=prim1[,2], filtered=track1[,2], denoised=sapply(dd1, function(x) sum(x$denoised)))
```

Sequence table:
```{r seqtab1}
st1 <- makeSequenceTable(dd1); dim(st1)
```
Save progress here before moving to PBB2.
```{r save seq table1}
saveRDS(st1, "RDS/PBB1_2.rds")
```


## PBB2

Remove primers and orient reads:
```{r primers2, message=FALSE, warning=FALSE}
nops2 <- file.path(pathPBB2, "noprimers", basename(fnsPBB2))
prim2 <- removePrimers(fnsPBB2, nops2, primer.fwd=B49873_e, primer.rev=dada2:::rc(A50272_f), orient=TRUE)
```

Inspect length distribution.
```{r length-distro}
lens.fn2 <- lapply(nops2, function(fn) nchar(getSequences(fn)))
lens2 <- do.call(c, lens.fn2)
hist(lens2, 100)
min(lens2)
max(lens2)
```

Filter and Trim
maxEE=8
```{r filter2}
filts2 <- file.path(pathPBB2, "noprimers", "filtered", basename(fnsPBB2))
track2 <- filterAndTrim(nops2, filts2, minQ=3, minLen=100, maxN=0, rm.phix=FALSE, maxEE=8)
track2
```


Dereplicate:
```{r derep2, message=FALSE}
drp2 <- derepFastq(filts2, verbose=TRUE)
```

Learn errors:
```{r learn-err2}
err2 <- learnErrors(drp2, errorEstimationFunction=PacBioErrfun, BAND_SIZE=32, multithread=TRUE)
saveRDS(err2, file.path(path.rds, "err_PBB2_2.rds"))
```

Inspect errors:
```{r see-err2}
plotErrors(err2)
```

Looks good.

Denoise:
```{r dada2-sample-2}
dd2 <- dada(drp2, err=err2, BAND_SIZE=32, multithread=TRUE)
saveRDS(dd2, file.path(path.rds, "PBB2_dd2_2.rds"))
```

Read tracking for PBB2:
```{r}
cbind(ccs=prim2[,1], primers=prim2[,2], filtered=track2[,2], denoised=sapply(dd2, function(x) sum(x$denoised)))
```

Sequence table:
```{r seqtab2}
st2 <- makeSequenceTable(dd2); dim(st2)
```
Save progress for PBB2 and next merge datafiles for downstream analyses.
```{r save seq table2}
saveRDS(st2, "RDS/PBB2_2.rds")
```


### Combine all quality-trimmed files

```{r combine-files}
st1 <- readRDS("RDS/PBB1_2.rds")
st2 <- readRDS("RDS/PBB2_2.rds")
st3 <- readRDS("RDS/PBB3_2.rds")
st.all <- mergeSequenceTables(st1, st2, st3)
```

### Remove chimeras, assign taxonomy and save progress
```{r remove-chimeras}
st.all_chimera_free <- removeBimeraDenovo(st.all, method="consensus", multithread=TRUE, verbose = TRUE)
```


```{r assign taxonomy}
# Assign taxonomy
tax <- assignTaxonomy(st.all, "./trnF_db11Feb22.fa", minBoot = 80, tryRC = TRUE, multithread=TRUE) # Slowest part
head(unname(tax))
```

```{r save-progrss-taxonomy}
# Write to disk
saveRDS(st.all, "RDS/PBB_all_runs_2.rds")
saveRDS(tax, "RDS/PBB_all_runs_tax_2.rds")
```

> NOTE: taxonomy will be improved later, using BLAST (nt database), but only after the ASV table is further cleaned up in Phyloseq

# Move analysis over to Phyloseq

## Load all runs RDS files (ASV table file and taxonomy file)
Also load metadata file, where sample names are row names

```{r phyloseq-load-data}
library(dplyr)
st.all_chimera_free <- readRDS("RDS/PBB_all_runs_2.rds")
tax <- readRDS("RDS/PBB_all_runs_tax_2.rds")
metadata <- read.table("./metadata_assembly/Burrower_bugs_metadata.csv", row.names = 3, sep = ",", header = TRUE) # Note, exclude first column (row number) before loading this spreadsheet
metadata$SequenceBatch <- as.factor(metadata$SequenceBatch)
metadata$collection_date <- as.Date(metadata$collection_date, format = "%m/%d/%y")
metadata$Month <- factor(metadata$Month, levels = month.name)
```

Make Phyloseq object
```{r phyloseq-make-object-1}
p_data <- phyloseq(otu_table(st.all_chimera_free, taxa_are_rows = FALSE), 
                   sample_data(metadata),
                   tax_table(tax))
```

Store DNA sequences in different slot and simplify ASV names
```{r phyloseq-format-data}
dna <- Biostrings::DNAStringSet(taxa_names(p_data))
names(dna) <- taxa_names(p_data)
p_data <- merge_phyloseq(p_data, dna)
taxa_names(p_data) <- paste0("ASV", seq(ntaxa(p_data)))
p_data
```

## Decontam

```{r decontam-eval-data}
library(decontam)
df <- as.data.frame(sample_data(p_data)) 
df$LibrarySize <- sample_sums(p_data)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
ggplot(data=df, aes(x=Index, y=LibrarySize, color=sample_type)) + geom_point() 
sample_data(p_data)$is.neg <- sample_data(p_data)$sample_type == "control" 

# adopting prevalence method to identify contaminants based on controls
# sensitivity set to low (threhold = 0.1)
contamdf.prev.01 <- isContaminant(p_data, method="prevalence", neg="is.neg") 
table(contamdf.prev.01$contaminant) 
contaminant.list <- which(contamdf.prev.01$contaminant)
tax_table(p_data)[contaminant.list,]

# sensitivity set to high (threshold = 0.5)
contamdf.prev.05 <- isContaminant(p_data, method="prevalence", neg="is.neg", threshold = 0.5) 
table(contamdf.prev.05$contaminant) 
contaminant.list <- which(contamdf.prev.05$contaminant)
tax_table(p_data)[contaminant.list,]
```

We identified via PCR that some reagents (e.g. some primer-barcode combinations) were likely contaminated with plant DNA. 
Suspect contaminants such as Citrus only show up as contaminant when sensitivity is increased to 0.5.

Let's inspect how contaminants prevalence is divided between control and samples

```{r decontam-eval-distribution}
# Make phyloseq object of presence-absence in negative controls and true samples
ps.pa <- transform_sample_counts(p_data, function(abund) 1*(abund>0))
ps.pa.neg <- prune_samples(sample_data(ps.pa)$sample_type == "control", ps.pa)
ps.pa.pos <- prune_samples(sample_data(ps.pa)$sample_type == "sample", ps.pa)
# Make data.frame of prevalence in positive and negative samples
# for threshold=0.1
df.pa1 <- data.frame(pa.pos=taxa_sums(ps.pa.pos), pa.neg=taxa_sums(ps.pa.neg),
                      contaminant=contamdf.prev.01$contaminant)
ggplot(data=df.pa1, aes(x=pa.neg, y=pa.pos, color=contaminant)) + geom_point() +
  xlab("Prevalence (Negative Controls)") + ylab("Prevalence (True Samples)") + ggtitle("Decontamination threshold=0.1")
# for threshold=0.5
df.pa5 <- data.frame(pa.pos=taxa_sums(ps.pa.pos), pa.neg=taxa_sums(ps.pa.neg),
                      contaminant=contamdf.prev.05$contaminant)
ggplot(data=df.pa5, aes(x=pa.neg, y=pa.pos, color=contaminant)) + geom_point() +
  xlab("Prevalence (Negative Controls)") + ylab("Prevalence (True Samples)") + ggtitle("Decontamination threshold=0.5")

p_data.ord <- ordinate(p_data, "PCoA", "bray")
plot_ordination(p_data, p_data.ord, type="sample", color="sample_type", shape="SequenceBatch")+ geom_point(size=3)
```

Because it is hard to distinguish whether contaminated primers were contaminated by the samples themselves, or whether it was contaminated water that was used for dilution, or both (likely), there is a risk that a more rigorous cutoff for contaminants will likely remove false-positives as well.
On the other hand, a soft filter may leave out some likely contaminants such as Citrus.
```{r decontam-clean-data}
p_data.noncontam <- prune_taxa(!contamdf.prev.05$contaminant, p_data)
p_data
p_data.noncontam 
# Remove control samples from analysis
p_data.clean <-subset_samples(p_data.noncontam, sample_type!="control")
# Remove samples with zero ASV (zero column sums)
p_data.clean <- subset_samples(p_data.clean, sample_sums(p_data.clean) > 0)
```


> At this point, the taxonomy table is likely to contain fewer (or no) spurious sequences
> Therefore, this is a good point to export the taxonomy table as a fasta file and run Blast to improve genus assignment (i.e. replace NAs)

#### To save taxonomy table use the following command

```{r save-taxonomy-table}
p_data.clean %>%
  tax_table() %>%
  write.csv("./PBB123_asv_11182024.csv")
```

#### To save sequences as a fasta file for blast, use the following command
```{r save-sequences-fasta}
p_data.clean %>%
      refseq() %>%
      Biostrings::writeXStringSet("./PBB123_asv_11182024.fasta", append=FALSE,
                                  compress=FALSE, compression_level=NA, format="fasta")
```

#### Save decontaminated phyloseq object and its components, so that you can pick up analysis from here after blast is done

```{r save-progress}
saveRDS(otu_table(p_data.clean), "RDS/p_data.cleanOTU.rds")
saveRDS(sample_data(p_data.clean), "RDS/p_data.cleanSampleData.rds")
saveRDS(tax_table(p_data.clean), "RDS/p_data.cleanOTUTaxaTable.rds")
saveRDS(refseq(p_data.clean), "RDS/p_data.cleanOTURefSeq.rds")
```



> This concludes part 1 of cleaning, clustering and classifying sequences in these samples. The next portion, part 2, is focused on determining taxonomy for some of the undetermined taxa, as well as excluding taxa that aren't relevant for this study, such as bryphoyte and algae.

# Part 2, adding results from blast search


Prepare environment:


Load Data

```{r phyloseq-load-improved-data}
p_data.cleanOTU <- readRDS("RDS/p_data.cleanOTU.rds")
p_data.cleanSampleData <- readRDS("RDS/p_data.cleanSampleData.rds")
p_data.cleanOTUTaxaTable <- readRDS("RDS/p_data.cleanOTUTaxaTable.rds")
p_data.cleanOTURefSeq <- readRDS("RDS/p_data.cleanOTURefSeq.rds")
```

Make Phyloseq object
```{r phyloseq-make-object}
p_data.clean <- phyloseq(otu_table(p_data.cleanOTU, taxa_are_rows = FALSE), 
                   sample_data(p_data.cleanSampleData),
                   tax_table(p_data.cleanOTUTaxaTable),
                   refseq(p_data.cleanOTURefSeq))
```


>Blast was executed using script under the file name blast_ASV_pbb123.sh
PBB123_asv_03292023.csv (taxonomy table) was edited and every NA was changed if taxonomic information was available (a blast hit result)
NOTE: taxonomical assignments ranged from about 95% to 100% and some NAs remained with no classification via Blast
#### Add new taxonomical information to Phyloseq object

```{r add-new-taxon-table}
new_tax_table <- read.csv("PBB123_asv_11182024.csv", row.names = 1)
new_tax <- tax_table(as.matrix(new_tax_table))
tax_table(p_data.clean) <- new_tax
```

## Exploratory analysis

List all instances where Genus == NA
```{r list-genus-NA}
#tax_table(p_data.clean) %>% as("matrix") %>% as_tibble 
p_data.clean %>%
  subset_taxa(is.na(Genus)) %>%
 tax_table()
```

All but a few instance of Genus=NA are ASV where Kingdom is also NA. As a conservative approach, I will exclude everything that is not identified as Viridiplantae at the Kingdom level

Remove singletons if needed and taxa where Order = NA
Remove instances where algae was identified
Remove samples where count is zero
```{r prepare-for-downstream-diversity-analysis}
tail(taxa_sums(p_data.clean) < 2)
p_data.clean <- prune_taxa(taxa_sums(p_data.clean) > 1, p_data.clean) %>% subset_taxa(., Kingdom == "Viridiplantae")
p_data.clean <- subset_samples(p_data.clean, sample_sums(p_data.clean) > 0)
```

Sanity check - is there any NA left for Genus?
```{r list-genus-NA-2}
#tax_table(p_data.clean) %>% as("matrix") %>% as_tibble 
p_data.clean %>%
  subset_taxa(is.na(Genus)) %>%
 tax_table()
```

These ASVs were found, after blasting again, to have indeed an inconclusive identification for Genus. In addition, there is taxa here that are still likely not part of these bugs' diet (e.g. Bryophytes).

Checking for mosses and other unexpected taxa
```{r are-there-mosses}
# are there unexpected taxa?
unique(tax_table(p_data.clean)[,3])
```

Mosses are the only unexpected taxa.
Remove mosses with the following command
```{r remove-mosses}
# How many ASVs are mossses?
p_data.clean %>% subset_taxa(., Class == "Bryopsida") %>% tax_table() # two ASVs
p_data.clean
p_data.clean <- subset_taxa(p_data.clean, Class != "Bryopsida")
p_data.clean # now there are two fewer ASVs
```


```{r save-progress}
saveRDS(otu_table(p_data.clean), "RDS/p_data.cleanOTU_cleaningcompleted.rds")
saveRDS(sample_data(p_data.clean), "RDS/p_data.cleanSampleData_cleaningcompleted.rds")
saveRDS(tax_table(p_data.clean), "RDS/p_data.cleanOTUTaxaTable_cleaningcompleted.rds")
saveRDS(refseq(p_data.clean), "RDS/p_data.cleanOTURefSeq_cleaningcompleted.rds")
```


>This concludes part 2 of (the bulk of) cleaning up and classifying data. Next, in part 3, I do more hypothesis-guided analyses, including figures and statistical tests.

#Part 3

Load Data

```{r phyloseq-load-improved-data}
p_data.cleanOTU <- readRDS("RDS/p_data.cleanOTU_cleaningcompleted.rds")
metadata <- readRDS("RDS/p_data.cleanSampleData_cleaningcompleted.rds")
p_data.cleanOTUTaxaTable <- readRDS("RDS/p_data.cleanOTUTaxaTable_cleaningcompleted.rds")
p_data.cleanOTURefSeq <- readRDS("RDS/p_data.cleanOTURefSeq_cleaningcompleted.rds")
```

Make Phyloseq object
```{r phyloseq-make-object}
p_data.clean <- phyloseq(otu_table(p_data.cleanOTU, taxa_are_rows = FALSE), 
                   sample_data(metadata),
                   tax_table(p_data.cleanOTUTaxaTable),
                   refseq(p_data.cleanOTURefSeq))
```


For plotting, scale counts to relative abundance
```{r scale_abundance}
p_data.clean_rel.ab <- transform_sample_counts(p_data.clean, function(x) 100*x/sum(x))
```

First I will find out the distribution of sample size
```{r sample-size-variation}
len_sample <- as.data.frame(sample_sums(p_data.clean))
median(len_sample$`sample_sums(p_data.clean)`)
mean(len_sample$`sample_sums(p_data.clean)`)
min(len_sample$`sample_sums(p_data.clean)`)
max(len_sample$`sample_sums(p_data.clean)`)
```

At this point I realized that since we have two different species in this dataset, it may make more sense to separate these species before rarefying
After looking at sample collection information, DNA extraction, number of samples left in the freezer, it was concluded that only samples in batches 2 and 3 (trnF data) were also submitted to RAD-seq
Therefore, this species separation excluded batch 1 data
```{r separate-species-2}
p_data.clean_batch1out <- subset_samples(p_data.clean, SequenceBatch != "1")
Pbilineatus <- subset_samples(p_data.clean_batch1out, Species=="Pbilineatus") %>% prune_taxa(taxa_sums(.) > 1, .)
Dlugubris <- subset_samples(p_data.clean_batch1out, Species=="Dlugubris") %>% prune_taxa(taxa_sums(.) > 1, .)
Pbilineatus
Dlugubris
```

Now, what is the sample size variation per species?
```{r sample-size-variation-2}
# P. bilineatus
len_samplepb <- as.data.frame(sample_sums(Pbilineatus))
median(len_samplepb$`sample_sums(Pbilineatus)`)
mean(len_samplepb$`sample_sums(Pbilineatus)`)
min(len_samplepb$`sample_sums(Pbilineatus)`)
max(len_samplepb$`sample_sums(Pbilineatus)`)
# D. lugubris
len_sampledl <- as.data.frame(sample_sums(Dlugubris))
median(len_sampledl$`sample_sums(Dlugubris)`)
mean(len_sampledl$`sample_sums(Dlugubris)`)
min(len_sampledl$`sample_sums(Dlugubris)`)
max(len_sampledl$`sample_sums(Dlugubris)`)
```

To visualize distribution in each case, I will exclude the largest samples and focus on the other samples

```{r plot-sample-size-pbilineatus}
#Exclude largest samples to build histogram
len_samplepb <-subset(len_samplepb, len_samplepb$`sample_sums(Pbilineatus)` < 1200)
median(len_samplepb$`sample_sums(Pbilineatus)`)
hist(len_samplepb$`sample_sums(Pbilineatus)`, 10)
```
```{r plot-sample-size-dlugubris}
#Exclude largest samples to build histogram
len_sampledl <-subset(len_sampledl, len_sampledl$`sample_sums(Dlugubris)` < 1200)
median(len_sampledl$`sample_sums(Dlugubris)`)
hist(len_sampledl$`sample_sums(Dlugubris)`, 10)
```

When extremely large samples are excluded, average number of reads per sample is under 100 reads per sample for either species.
Increasing maxEE to 8 did not make a substantial change - the results here are very similar to the results obtained earlier with maxEE=2.
It seems reasonable to use the cutoff for rarefying samples at 100 reads per sample, hopefully keeping close to half of the samples for each species. Since the cutoff is the same, then I can rarefy the p_data.clean, to run statistics to compare both species directly. If necessary, I can separate samples as well after they are rarefied.


This time I will rarefy samples to 100 reads per sample.

```{r rarefy-data-pbilineatus}
p_data.clean_rarefied <- rarefy_even_depth(p_data.clean, sample.size = 100, 
                                           verbose = FALSE, replace = TRUE)
p_data.clean
p_data.clean_rarefied
```


let's see what are the top 10 taxa
```{r taxa-relative-abundance}
relative_taxa_abundance <- (taxa_sums(p_data.clean_rarefied))/sum(taxa_sums(p_data.clean_rarefied))
sort(relative_taxa_abundance,TRUE)[1:10]
# When does relative abundance cross 90% ?
sum(sort(relative_taxa_abundance,TRUE)[1:10]) # 53%
sum(sort(relative_taxa_abundance,TRUE)[1:40]) # 87%
sum(sort(relative_taxa_abundance,TRUE)[1:50]) # 92%
sum(sort(relative_taxa_abundance,TRUE)[1:45]) # 90%
```
45 ASVs out of 129 correspond to 90% of all counts across all samples. We will focus on these 45 ASV.


# Visualizing Patterns

To facilitate visualizing, I will agglomerate all these ASV to the level of Family.

```{r plot-barplot-top47-Site}
top45 <- prune_taxa(names(sort(taxa_sums(p_data.clean_rarefied),TRUE)[1:45]), p_data.clean_rarefied)
p <- plot_bar(top45, fill = "Family") + facet_wrap(~Site, scales="free_x", nrow=1) + theme(axis.text.x = element_blank())
p
# Save image in PDF
pdf("Top45_per_Site_PBB123.pdf",height=12,width=21)
p
dev.off()
```

```{r plot-barplot-top47-Site-genus}
top45 <- prune_taxa(names(sort(taxa_sums(p_data.clean_rarefied),TRUE)[1:45]), p_data.clean_rarefied)
p <- plot_bar(top45, fill = "Genus") + facet_wrap(~Site, scales="free_x", nrow=1) + theme(axis.text.x = element_blank())
p
# Save image in PDF
pdf("Top45_per_Site_PBB_genus.pdf",height=12,width=21)
p
dev.off()
```

Remove Colony (control)
```{r plot-barplot-remove-Colony}
# Remove control samples from analysis
p_data.clean_Colony_out <-subset_samples(p_data.clean_rarefied, Site!="Colony")
p_data.clean_Colony_out_no_rarefy <- subset_samples(p_data.clean, Site!="Colony")
```

Questions we have to answer

(1) There are two species in this dataset, so the first thing we should do is separate them
(2) Second, do they both eat peanuts?
(3) Third, explore what explains their diet


```{r separate-species}
batches2and3 <- subset_samples(p_data.clean_Colony_out, SequenceBatch != "1")
# included a step to remove taxa that may become singletons when sequence batch 1 is removed
Pbilineatus <- subset_samples(batches2and3, Species == "Pbilineatus") %>% prune_taxa(taxa_sums(.) > 1, .) 
Dlugubris <- subset_samples(batches2and3, Species == "Dlugubris") %>% prune_taxa(taxa_sums(.) > 1, .)
```

Plot Peanut only
```{r plot-barplot-Peanut-only}
plot_peanuts <- subset_taxa(p_data.clean_Colony_out, Genus=="Arachis")
p <- plot_bar(plot_peanuts, fill = "Genus") + facet_wrap(~Site, scales="free_x", nrow=1) + theme(axis.text.x = element_blank())
p
# Save image in PDF
pdf("Peanut_content_per_Site_PBB.pdf",height=12,width=21)
p
dev.off()

```

Plot Peanut for each species
```{r plot-barplot-Peanut-only-2}
plot_peanuts <- subset_taxa(p_data.clean_Colony_out, Genus=="Arachis") %>% subset_samples(., SequenceBatch!="1")
p <- plot_bar(plot_peanuts, fill = "Genus") + facet_wrap(~sample_data(plot_peanuts)$Species, scales="free_x", nrow=1) + theme(axis.text.x = element_blank())
p
# Save image in PDF
pdf("Arachis_detected_in_each_species.pdf",height=12,width=21)
dev.off()
```

The plot above suggests that Dlugubris does not consume peanut. However, except for one sample, Pbilineatus was also negative for Arachis. Because of all the filtering and losses due to rarefying data, it might be worth to search for Arachis at samples before rarefying data

```{r peanuts-before-rarefying}
plot_peanuts_Dlugubris <- subset_taxa(p_data.clean_Colony_out_no_rarefy, Genus=="Arachis") %>% subset_samples(., SequenceBatch!="1") %>% subset_samples(., Species=="Dlugubris")
plot_peanuts_Pbilineatus <- subset_taxa(p_data.clean_Colony_out_no_rarefy, Genus=="Arachis") %>% subset_samples(., SequenceBatch!="1") %>% subset_samples(., Species=="Pbilineatus")
p1 <- plot_bar(plot_peanuts_Dlugubris, fill = "Genus") + theme(axis.text.x = element_blank()) + ggtitle("Dlugubris")
p2 <- plot_bar(plot_peanuts_Pbilineatus, fill = "Genus") + theme(axis.text.x = element_blank()) + ggtitle("Pbilineatus")
p1
p2
# Save image in PDF
#pdf("Arachis_detected_in_each_species_before_rarefying_data.pdf",height=12,width=21)
#p
#dev.off()
```


```{r plot-barplot-Peanuts-gut-content-Season}
p1 <- plot_bar(plot_peanuts_Dlugubris, fill = "Genus") + facet_wrap(~Season, scales="free_x", nrow=1) + theme(axis.text.x = element_blank()) + ggtitle("Dlugubris")
p2 <- plot_bar(plot_peanuts_Pbilineatus, fill = "Genus") + facet_wrap(~Season, scales="free_x", nrow=1) + theme(axis.text.x = element_blank()) + ggtitle("Pbilineatus")
p1
p2
# Save image in PDF
#pdf("Peanut_content_per_Site_PBB.pdf",height=12,width=21)
#p
#dev.off()

```

It is clear that peanut is consumed by Dlugubris. However, we don't have sampling of Dlugubris for Spring at this point.
Interestingly, peanuts are planted from April-May, and they are harvested 140-150 days after planting, that is between August and September/October, which includes all the three seasons we sampled (Spring, Summer, Fall) - source: National Peanut Board (https://www.nationalpeanutboard.org)

```{r plot-richness-by-species-rarefied-data}
p1 <- plot_richness(Dlugubris, x="Month", color="Site", measures=c("Observed")) + ggtitle("Dlugubris") + geom_jitter()
p1
p2 <- plot_richness(Pbilineatus, x="Month", color="Site", measures=c("Observed")) + ggtitle("Pbilineatus") + geom_jitter()
p2
#pdf("richness_PBB.pdf",height=12,width=21)
#p
#dev.off()
```

```{r ordination}
# For all data
p_data.c_r_ord <- ordinate(p_data.clean_Colony_out, "PCoA", "bray")
p_data.c_r_dist <- phyloseq::distance(p_data.clean_Colony_out, method = "bray")
p1 <- plot_ordination(p_data.clean_Colony_out, p_data.c_r_ord, type="sample", color="Season", shape="Species", title = "All data")+ geom_point(size=5)
p1

# For data only identified as either Pbilineatus or Dlugubris
filt_data <- subset_samples(p_data.clean_Colony_out, SequenceBatch != "1") %>% prune_taxa(taxa_sums(.) > 1, .)
filt_data_ord <- ordinate(filt_data, "PCoA", "bray")
filt_data_dist <- phyloseq::distance(filt_data, method = "bray")
p2 <- plot_ordination(filt_data, filt_data_ord, type="sample", color="Season", shape="Species", title = "Only Dlugubris and Pbilineatus")+ geom_point(size=5)
p2

# Save images in PDF
pdf("PCA_PBB_all.pdf",height=12,width=21)
p1
dev.off()
pdf("PCA_PBB_batches2and3.pdf",height=12,width=21)
p2
dev.off()
```

```{r finding core gut content for Pbilineatus}
library(metagMisc)
# Following taxa shared by 50% of Pbilineatus samples, for all seasons and then per season
core_Pbilineatus <- phyloseq_filter_prevalence(Pbilineatus, prev.trh = 0.5, abund.trh = NULL)
tax_table(core_Pbilineatus)
subset_samples(core_Pbilineatus, sample_sums(core_Pbilineatus) > 0)
core_PbilineatusSpring <- subset_samples(Pbilineatus, Season == "Spring") %>% phyloseq_filter_prevalence(., prev.trh = 0.5, abund.trh = NULL) # 4 samples
tax_table(core_PbilineatusSpring)
subset_samples(core_PbilineatusSpring, sample_sums(core_PbilineatusSpring) > 0)
core_PbilineatusSummer <- subset_samples(Pbilineatus, Season == "Summer") %>% phyloseq_filter_prevalence(., prev.trh = 0.5, abund.trh = NULL) # 6 samples
tax_table(core_PbilineatusSummer)
subset_samples(core_PbilineatusSummer, sample_sums(core_PbilineatusSummer) > 0)
core_PbilineatusFall <- subset_samples(Pbilineatus, Season == "Fall") %>% phyloseq_filter_prevalence(., prev.trh = 0.4, abund.trh = NULL) # 5 samples
tax_table(core_PbilineatusFall)
subset_samples(core_PbilineatusFall, sample_sums(core_PbilineatusFall) > 0)
```

```{r finding core gut content for Dlugubris}
# Following, taxa shared by 20% of the samples for Dlugubris, in general and per season
# Using 20% because above this threshold no ASV is found
core_Dlugubris <- phyloseq_filter_prevalence(Dlugubris, prev.trh = 0.2, abund.trh = NULL) #33 samples total
tax_table(core_Dlugubris)
subset_samples(core_Dlugubris, sample_sums(core_Dlugubris) > 0) #8 samples left
core_DlugubrisSummer <- subset_samples(Dlugubris, Season == "Summer") %>% phyloseq_filter_prevalence(., prev.trh = 0.2, abund.trh = NULL) # 7 samples
tax_table(core_DlugubrisSummer)
subset_samples(core_DlugubrisSummer, sample_sums(core_DlugubrisSummer) > 0) # 3 samples left
core_DlugubrisFall <- subset_samples(Dlugubris, Season == "Fall") %>% phyloseq_filter_prevalence(., prev.trh = 0.2, abund.trh = NULL) # 26 samples
tax_table(core_DlugubrisFall)
subset_samples(core_DlugubrisFall, sample_sums(core_DlugubrisFall) > 0) # 7 samples left
```

```{r permanova, message=FALSE}
Site <- sample_data(p_data.clean_Colony_out)$Site
Season <- sample_data(p_data.clean_Colony_out)$Season
SequenceBatch <- sample_data(p_data.clean_Colony_out)$SequenceBatch

library(vegan)
adonis2(p_data.c_r_dist ~ SequenceBatch) #Sequence Batch has an effect!
adonis2(p_data.c_r_dist ~ Site*Season, strata =SequenceBatch)

# Comparing species
Site <- sample_data(filt_data)$Site
Season <- sample_data(filt_data)$Season
SequenceBatch <- sample_data(filt_data)$SequenceBatch
SampleSpecies <- sample_data(filt_data)$Species

adonis2(filt_data_dist ~ SequenceBatch) #There is a batch effect!
adonis2(filt_data_dist ~ Season+SampleSpecies*Site, strata = SequenceBatch)

# If we exclude samples from Spring
Summer_Fall_only <- subset_samples(filt_data, Season!="Spring")
Summer_Fall_only_dist <- phyloseq::distance(Summer_Fall_only, method = "bray")

Site <- sample_data(Summer_Fall_only)$Site
Season <- sample_data(Summer_Fall_only)$Season
SequenceBatch <- sample_data(Summer_Fall_only)$SequenceBatch
SampleSpecies <- sample_data(Summer_Fall_only)$Species
adonis2(Summer_Fall_only_dist ~ SequenceBatch) # No effect of Sequencing Batch - could it be a cofounding variable with Season?
adonis2(Summer_Fall_only_dist ~ Season+SampleSpecies*Site)
adonis2(Summer_Fall_only_dist ~ SampleSpecies*Season)
adonis2(Summer_Fall_only_dist ~ SampleSpecies, strata = Season)

# If we look at data only for batch 3 (where we are 100% sure the samples are the same between trnf data and RAD-seq)
batch3_only <- subset_samples(p_data.clean_Colony_out, SequenceBatch == "3")
batch3_only_dist <- phyloseq::distance(batch3_only, method = "bray")

Site <- sample_data(batch3_only)$Site
Season <- sample_data(batch3_only)$Season
#SequenceBatch <- sample_data(batch3_only)$SequenceBatch
SampleSpecies <- sample_data(batch3_only)$Species
#adonis(batch3_only_dist ~ SequenceBatch)
adonis2(batch3_only_dist ~ Site+SampleSpecies+Season)
adonis2(batch3_only_dist ~ SampleSpecies, strata = Season)
```

Save all files for future reference
```{r eval=FALSE, saving-otu_table-taxonomy_table-metadata_table-and-fasta-files}
#rarefied dataset (Colony samples excluded)
otu_table(p_data.clean_rarefied) %>% as("matrix") %>% as_tibble %>% write.csv("./rarefied_otu_table.csv")
sample_data(p_data.clean_rarefied) %>% as("matrix") %>% as_tibble %>% write.csv("./rarefied_metadata_table.csv")
tax_table(p_data.clean_rarefied) %>% as("matrix") %>% as_tibble %>% write.csv("./rarefied_tax_table.csv")
refseq(p_data.clean_rarefied) %>%
      Biostrings::writeXStringSet("./PBB123_rarefied_03302023.fasta", append=FALSE,
                                  compress=FALSE, compression_level=NA, format="fasta")
#data not rarefied with Colony samples included
otu_table(p_data.clean)  %>% as("matrix") %>% as_tibble %>% write.csv("./otu_table03302023.csv")
sample_data(p_data.clean) %>% as("matrix") %>% as_tibble %>% write.csv("./metadata_table03302023.csv")
tax_table(p_data.clean) %>% as("matrix") %>% as_tibble %>% write.csv("./tax_table03302023.csv")
refseq(p_data.clean) %>%
      Biostrings::writeXStringSet("./PBB123_03302023_asv.fasta", append=FALSE,
                                  compress=FALSE, compression_level=NA, format="fasta")
```

```{r export-summary-table-as-per-ycl6-function, warning=FALSE, message=FALSE}
library("tidyverse")
# for rarefied data
write.table(p_data.clean_rarefied %>%
        psmelt() %>% arrange(OTU) %>% rename(ASV = OTU) %>%
        select(ASV, Kingdom, Phylum, Class, Order, Family, Genus, Sample, Abundance) %>% spread(Sample, Abundance), 
	file = "summary_pbb123-rarefied03302023.tsv", sep = "\t", quote = F, row.names = F, col.names = T)
write.table(Pbilineatus %>%
        psmelt() %>% arrange(OTU) %>% rename(ASV = OTU) %>%
        select(ASV, Kingdom, Phylum, Class, Order, Family, Genus, Sample, Abundance) %>% spread(Sample, Abundance), 
	file = "summary_Plineatus-rarefied03302023.tsv", sep = "\t", quote = F, row.names = F, col.names = T)
write.table(Dlugubris %>%
        psmelt() %>% arrange(OTU) %>% rename(ASV = OTU) %>%
        select(ASV, Kingdom, Phylum, Class, Order, Family, Genus, Sample, Abundance) %>% spread(Sample, Abundance), 
	file = "summary_Dlugubris-rarefied03302023.tsv", sep = "\t", quote = F, row.names = F, col.names = T)
# for non-rarefied data
write.table(p_data.clean %>% 
        psmelt() %>% arrange(OTU) %>% rename(ASV = OTU) %>%
        select(ASV, Kingdom, Phylum, Class, Order, Family, Genus, Sample, Abundance) %>% spread(Sample, Abundance), 
	file = "summary_pbb03302023.tsv", sep = "\t", quote = F, row.names = F, col.names = T)
write.table(subset_samples(p_data.clean, SequenceBatch !=1 & Species == "Pbilineatus") %>%
        psmelt() %>% arrange(OTU) %>% rename(ASV = OTU) %>%
        select(ASV, Kingdom, Phylum, Class, Order, Family, Genus, Sample, Abundance) %>% spread(Sample, Abundance), 
	file = "summary_Plineatus03302023.tsv", sep = "\t", quote = F, row.names = F, col.names = T)
write.table(subset_samples(p_data.clean, SequenceBatch !=1 & Species == "Dlugubris") %>%
        psmelt() %>% arrange(OTU) %>% rename(ASV = OTU) %>%
        select(ASV, Kingdom, Phylum, Class, Order, Family, Genus, Sample, Abundance) %>% spread(Sample, Abundance), 
	file = "summary_Dlugubris03302023.tsv", sep = "\t", quote = F, row.names = F, col.names = T)
```

```{r include=FALSE, eval=FALSE}
write.table(p_data.clean_Colony_out_no_rarefy %>% tax_glom(taxrank = "Genus") %>% 
        transform_sample_counts(function(x) {x/sum(x)}) %>% psmelt() %>% 
        select(Genus, Sample, Abundance) %>% spread(Sample, Abundance), 
	file = "pbb_clean_norarefy.relative_abundance.genus.tsv", sep = "\t", quote = F, row.names = F, col.names = T)

```


